platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/hash/VirtualHasher.java

```diff
@@ -198,62 +141,189 @@ public Hash hash(
         return hash(hashReader, sortedDirtyLeaves, firstLeafPath, lastLeafPath, null);
     }
 
-    /**
-     * Hash the given dirty leaves and the minimal subset of the tree necessary to produce a single root hash.
-     * The root hash is returned.
-     *
-     * @param hashReader
-     * 		Return a {@link Hash} by path. Used when this method needs to look up clean nodes.
-     * @param sortedDirtyLeaves
-     * 		A stream of dirty leaves sorted in <strong>ASCENDING PATH ORDER</strong>, such that path
-     * 		1234 comes before 1235. If null or empty, a null hash result is returned.
-     * @param firstLeafPath
-     * 		The firstLeafPath of the tree that is being hashed. If &lt; 1, then a null hash result is returned.
-     * 		No leaf in {@code sortedDirtyLeaves} may have a path less than {@code firstLeafPath}.
-     * @param lastLeafPath
-     * 		The lastLeafPath of the tree that is being hashed. If &lt; 1, then a null hash result is returned.
-     * 		No leaf in {@code sortedDirtyLeaves} may have a path greater than {@code lastLeafPath}.
-     * @param listener
-     * 		A {@link VirtualHashListener} that will receive notification of all hashing events. Can be null.
-     * @return The hash of the root of the tree
-     */
+    static class BaseTask extends ForkJoinTask<Void> {
+
+        protected final AtomicInteger count;
+
+        public BaseTask(int n) {
+            this.count = new AtomicInteger(n);
+        }
+
+        @Override
+        public Void getRawResult() {
+            return null;
+        }
+
+        @Override
+        protected void setRawResult(Void value) {
+            // not used
+        }
+
+        @Override
+        protected boolean exec() {
+            return true;
+        }
+
+        void push() {
+            if (count.decrementAndGet() == 0) {
+                forkIt();
+            }
+        }
+
+        void push(final int p) {
+            if (count.addAndGet(-p) == 0) {
+                forkIt();
+            }
+        }
+
+        private void forkIt() {
+            if (Thread.currentThread() instanceof ForkJoinWorkerThread) {
+                fork();
+            } else {
+                pool.execute(this);
+            }
+        }
+    }
+
+    class ChunkHashTask extends BaseTask {
+
+        private final long path;
+
+        private final int height; // 1 for 3-node chunk, 2 for 7-node chunk, and so on
+
+        private ChunkHashTask out;
+
+        private final Hash[] ins;
+
+        private VirtualLeafRecord<K, V> leaf;
+
+        // isLeaf is just an optimization: if the task is known to be used as a leaf task,
+        // there is no need to allocate an array of hashes that will always be empty
+        ChunkHashTask(long path, int height, boolean isLeaf) {
+            super(1 + (1 << height));
```

Comment:
We can set the number of dependencies conditionally:
```
            super(1 + (isLeaf ? 1 : (1 << height)));
```
and use `push()` in `setLeaf(leaf)` eliminating a need in `push(p)`.

Reply:
`push(p)` is now eliminated.